File: PGE381L-AdvPetrophysics/Resources/Examples/Spatial Data Analysis/Code01_Cov_Corr_Variogram_Syn_Examples.m
%% This code calculates autocovarianve, autocorrelation, and variogram
% Example: Formations A, B, and C 

close all
clear all
clc

File = xlsread('ForA.xlsx');
% File = xlsread('ForB.xlsx');
% File = xlsread('ForC.xlsx');

Data = File(:,2);
h = File(:,1);

nLags = round(0.6*length(h));

for i = 0:nLags-1,
    Vario(i+1) = sum((Data(1:end-i)-Data(1+i:end)).^2)/(2*length(Data(1:end-i)));
    Cov(i+1) = sum((Data(1:end-i) - mean(Data)).*(Data(1+i:end)-mean(Data)))/(length(Data(1:end-i))-1);
end


%% Plot Covarianve and Correlation
figure;
grid on;box on;hold on
set(gcf,'color','white')

subplot(3,1,1)
stem(h,Data)
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('Calcite Concentration','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

subplot(3,1,2)
plot(h(1:nLags)-h(1),Cov,'color',[0.1 0.1 1],'Linewidth',2)
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('Cov(h)','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

subplot(3,1,3)
plot(h(1:nLags)-h(1),Cov/Cov(1),'color',[0.1 0.7 0.1],'Linewidth',2)
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('Corr(h)','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

%% Plot Variogram
figure;
grid on;box on;hold on
set(gcf,'color','white')

subplot(3,1,1)
stem(h,Data)
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('Calcite Concentration','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

subplot(3,1,2)
plot(h(1:nLags)-h(1),Cov,'color',[0.1 0.1 1],'Linewidth',2)
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('Cov(h)','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

subplot(3,1,3)
plot(h(1:nLags)-h(1),Vario,'color',[1 0.1 0.1],'Linewidth',2)
hold on
plot(h(1:nLags),Cov(1)-Cov,'r--')
xlabel('h','fontsize',16,'fontweight','bold');
ylabel('\gamma(h)','fontsize',16,'fontweight','bold');
set(gca,'FontWeight','bold','FontSize',16,'LineWidth',2);

File: PGE381L-AdvPetrophysics/Resources/Examples/Spatial Data Analysis/Python Codes/Code01_Cov_Corr_Variogram_Syn_Examples.py
## This code calculates autocovarianve, autocorrelation, and variogram
# Example: Formations A, B, and C

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Read file and load data to variable df
fileName = 'formationA.xlsx';
#fileName = 'formationB.xlsx';
#fileName = 'formationC.xlsx';

df = pd.read_excel(fileName, sheet_name='Sheet1')

# Print headings available in the file
print("Column headings:")
headings=df.columns

# Load data of the columns in different arrays.
Data = df[headings[1]].values
h = df[headings[0]].values

nLags = round(0.6*len(h));

Vario = np.zeros(nLags)
Cov = np.zeros(nLags)
end = len(Data)
for i in range(0,nLags):
   pato1 = Data[0:end-i]
   pato2 = Data[i:end]
   lenPato = len(Data[0:-(1+i)])
   Vario[i] = sum((Data[0:end-i]-Data[i:end])**2)/(2*len(Data[0:end-i]));
   Cov[i] = sum((Data[0:end-i] - np.mean(Data))*(Data[i:end]-np.mean(Data)))/(len(Data[0:end-i])-1);

#%% Plot Covarianve and Correlation
   
plt.figure(figsize=(10,7))
plt.subplot(3, 1, 1)
plt.stem(h,Data,linefmt='b-', markerfmt='bo');
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('Calcite Concentration', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.subplot(3, 1, 2)
plt.plot(h[0:nLags]-h[0],Cov,color=[0.1,0.1,1],linewidth=2)
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('Cov(h)', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.subplot(3, 1, 3)
plt.plot(h[0:nLags]-h[0],Cov/Cov[0],color=[0.1,0.7,0.1],linewidth=2)
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('Corr(h)', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.tight_layout()

#%% Plot Variogram

plt.figure(figsize=(10,7))
plt.subplot(3, 1, 1)
plt.stem(h,Data, linefmt='b-', markerfmt='bo')
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('Calcite Concentration', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.subplot(3, 1, 2)
plt.plot(h[0:nLags]-h[0],Cov,'b',linewidth=2)
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('Cov(h)', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.subplot(3, 1, 3)
plt.plot(h[0:nLags]-h[0],Vario,'r',linewidth=2)
plt.plot(h[0:nLags],Cov[0]-Cov,'r--')
plt.xlabel('h', fontsize=16, weight='bold')
plt.ylabel('$\gamma$(h)', fontsize=16, weight='bold')
plt.xticks(fontsize=16, weight='bold')
plt.yticks(fontsize=16, weight='bold')
plt.grid(zorder=0, linestyle='--')

plt.tight_layout()
plt.show()
